Added connection-specific interfaces and factory functions for exporting instead

microbit-grace opened the pull request:
- Added specific interfaces for different connection types. This is helpful for mocking.
- Exported new functions for creating the connections instead of the class.
- Updated demo and docs to match changes.

microbit-grace commented on lib/device.ts:
```diff
@@ -242,5 +234,40 @@ export interface DeviceConnection
+  flash(dataSource: FlashDataSource, options: {}): Promise<void>;
+}
+
+export interface DeviceWebBluetoothConnection extends DeviceConnection {
+  /**
+   * Sets micro:bit name filter for device requesting.
+   *
+   * @param name The name of the micro:bit.
+   */
+  setNameFilter(name: string): void;
+}
```
Maybe I should extend this interface to other methods that are not to with the connection flow (e.g. `getAccelerometerData`, `getAccelerometerPeriod` ...etc) so that others using these methods can mock them as well?

microbit-matt-hillsdon commented on src/demo.ts:
```diff
@@ -194,7 +195,7 @@ const createConnectSection = (type: ConnectionType): Section => {
 
 const createFlashSection = (): Section => {
-  if (!connection.flash) {
+  if (typeof connection["flash"] !== "function") {
```
Let's use ConnectionType to determine this.

microbit-grace replied:
Have address this here: https://github.com/microbit-foundation/microbit-connection/pull/50/commits/29257d66b2cf2ea80286b7a1174aacd513538c0c

microbit-grace commented:
See below for commits addressing the comments:

> * Rename the implementations to e.g. MicrobitWebUSBConnectionImpl and the interfaces to e.g. MicrobitWebUSBConnection

[96b4af3](https://github.com/microbit-foundation/microbit-connection/pull/50/commits/96b4af34ed0f4aacb5925315f92806b798282f83), [9ab8849](https://github.com/microbit-foundation/microbit-connection/pull/50/commits/9ab88495440b4462a472591a7742debbaa025e4d)

> * Move the interfaces to the relevant files (e.g. usb.ts).

[1d47926](https://github.com/microbit-foundation/microbit-connection/pull/50/commits/1d47926d3199ad6269e3fd1c7bd65416e31e78b9)

> * Stop exporting the implementation classes by introducing createWebUSBConnection(options) function and similar and exporting that instead 

[c2ab488](https://github.com/microbit-foundation/microbit-connection/pull/50/commits/c2ab4882e83c80f2e602fb0a15086e1069f7c275), [29257d6](https://github.com/microbit-foundation/microbit-connection/pull/50/commits/29257d66b2cf2ea80286b7a1174aacd513538c0c)
 
> Please can we also look at whether there's a better approach to typing the events? Ideally we'd be able to add common events only via DeviceConnection and events that only applied e.g. USB would only be available via that interface.

Maybe like this [e5929d2](https://github.com/microbit-foundation/microbit-connection/pull/50/commits/e5929d289b4de1509b8d856582d6d78ea3ca48a5)?

Have also updated the documentation here: [65e8de4](https://github.com/microbit-foundation/microbit-connection/pull/50/commits/65e8de4cb06888cce5936e0718b1bc7e4a4c75fe)

microbit-matt-hillsdon commented:
Code changes look good. I'd like to spend a bit of time playing around with how it feels to use before we go ahead.

microbit-robert reviewed (approved):
I think this approach makes sense. Looks good.
